### DISCUSSION TITLE
How to refer to current object on migration when fill_expr>

### DISCUSSION THREAD

**Discussion by riolly** [geldata/gel#4072] [Q&A]
I try to migrate the property of subtype to supertype like this

Before
```
abstract type Person {
  property name -> str
}

type PC extending Person {
  required property first_appearance -> cal::local_date;
  property last_appearance -> cal::local_date;
}

type NPC extending Person {}
  
type Vampire extending Person {}
```

Onto
```
abstract type Person {
  property name -> str
  property last_appearance -> cal::local_date;
  required property first_appearance -> cal::local_date;
}

type PC extending Person {
}
...
```

And I want to set first_appearance property of Person with .first_appearance when the type is NPC and '1885-11-02' for non-NPC.
```
Please specify an expression to populate existing objects in order to make property 'first_appearance' of object type 'default::Person' required:
fill_expr> 
<cal::local_date>'1885-11-02' if not (Person is PC) else [is PC].first_appearance
```

but it throws an error 
`error: SchemaError: possibly more than one element returned by the USING clause for the alteration of property 'first_appearance' of object type 'default::Townsperson', while a singleton is expected`

I also try to use `assert_single` but when running the migration it returns an error
`edgedb error: CardinalityViolationError: assert_single violation: more than one element returned by an expression`

I am confused because when I do this query
```
select Person {
  name,
  first_appearance := 
    <cal::local_date>'1885-11-02' if not (Person is PC) 
    else [is PC].first_appearance
  };
```
Everything work Just fine

So why does it return more than one element?

**Comment by riolly** [comment]
I also try this (which I guess I have to make a query first for the person??)
```
fill_expr>
  with
    person := (select Person filter Person.id = .id)
  select 
    <cal::local_date>'1885-02-02' if not (person is PC) 
    else person[is PC].first_appearance

```

but it produces this error
```
edgedb error: InternalServerError: column Person~3.2580d1e7-0333-11ed-9d38-138989de27dd does not exist
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      edb.errors.InternalServerError: column Person~3.2580d1e7-0333-11ed-9d38-138989de27dd does not exist
```

edgedb v1.4
edgedb-cli v1.2

**Reply by elprans** [reply]
The cardinality violation is because `Person` in your original `fill_expr` refers to a set of _all_ persons, i.e. it's not bound to the "current" person.  To refer to the "current" object use the special `__source__` variable, e.g.:

```
fill_expr> '1885-11-02' if __source__ is not PC else [is PC].first_appearance
```

Though, since the `[is PC]` type intersection would produce an empty set on non-PC objects, a simpler expression could be a coalesce:

```
fill_expr> [is PC].first_appearance ?? '1885-11-02'
```

Though this would use the default date if `first_appearance` is unset, which might or might not be what you want.

**Reply by riolly** [reply]
Thank you so much
